### PR TITLE
[expo-updates] send expo-current-update-id header

### PR DESCRIPTION
**Pull request by esamelson**
# Why

Allow metering for EAS updates, cc @jkhales -- by passing the current update ID with manifest requests, we can make a pretty good guess on the service side about # of update downloads by comparing this to the served update ID.

# How

Pass the launched update through to the manifest request and include its ID in the `expo-current-update-id` header.

# Test Plan

Add logging to determine that this codepath is being triggered. Tested in managed workflow in both Android and iOS clients, automatic & manual (JS-triggered) manifest checks.

Added a few simple unit tests as well.

### CODE REVIEW

**Review by jkhales — APPROVED** [review]
This required a lot more code modification than I realized! Thanks for doing this! 🙏

**Review by ide — COMMENTED** [review]
Is the goal to roughly guess when a client will download a whole update? It seems like we need to handle these cases:

Update is fetched:
  - There is a new update. App checks for update at launch, and the server responds with an update that the app doesn't already have 
  - There is a new update. App runs an update that calls `Updates.fetchUpdateAsync()` from JS and downloads a new update from the server

Update is not fetched:
- There is no new update. App checks for update at launch or JS calls `Updates.fetchUpdateAsync()`. Either way, there is nothing new to download.
- There is a new update. App calls `Updates.checkForUpdatesAsync()` and finds an update but does not download it except for the manifest.

If I understand correctly, this commit sends Expo-Current-Update-ID only once an update is running and the JS API is called. Additionally, if someone checks for an update from JS but doesn't download it, it seems like that would count towards a manifest download as well.

Would we use Expo-Current-Update-ID as confirmation that an app has downloaded the specified update? It seems like we'd need to "lag" behind by one update:

- App has update 1
- App checks for new update, sends ECUI: 1
- Server sends update 2's manifest. Do not assume the app downloaded update 2 even though it might actually be happening at this moment.
- Time passes
- App checks for new update, sends ECUI: 2
- Server knows app has updated to update 2 sometime between now and when it first sent down update 2's manifest

Basically: is it correct to say that we only know that an app has updated to update N when it is looking for update N+1, not when it actually downloads update N?

**Comment by esamelson** [comment]
> If I understand correctly, this commit sends Expo-Current-Update-ID only once an update is running and the JS API is called.

This is not quite correct. This header is sent in the initial manifest request, too, and the value is the ID of the latest update that exists on disk -- basically, the update that would be launched if the timer were to run out (or, if the timer has already run out, the update that has launched/is currently launching).

So by sending this header and being able to compare on the server if the ID matches the one it's responding with, we handle the first 3 of your 4 bullet points.

> - There is a new update. App calls `Updates.checkForUpdatesAsync()` and finds an update but does not download it except for the manifest.

In the current implementation, we might overcount here if the user repeatedly calls `Updates.checkForUpdateAsync()` and there is a new update, but it isn't downloaded. We could avoid this by specifically not sending this header in requests from `Updates.checkForUpdateAsync()` -- @jkhales does that seem reasonable to you?

> Would we use Expo-Current-Update-ID as confirmation that an app has downloaded the specified update?

Will defer to @jkhales on this question. While this seems more correct, seems like it might require keeping some map on the server side of device IDs to update IDs, which is potentially problematic in terms of privacy?

**Comment by ide** [comment]
Thanks for the clarification. I mistakenly assumed that "launchedUpdate" referred to an already-running update and didn't apply to the scenario where the app checks for an update upon cold launching.

I wonder if it'd make sense actually to send an extra header that tells whether the client is checking for an update upon launch or on demand (via the JS API). One benefit of this is that `Updates.checkForUpdateAsync()` could then send the `Expo-Current-Update-ID` header (as it currently does in this PR) for consistency. One quick idea: `Expo-Update-Check-Reason: launch | app | ... (periodic background checks, silent push notifications, etc...)`

**Comment by ide** [comment]
Eric and I chatted a bit today about thinking through a wider set of potential use cases to avoid painting ourselves into a corner with a proposal that's too inflexible or fragile. There are cases we need to consider like embedded updates that are newer than published updates (different ID, newer commit time), plus the fact that client fields are inherently untrustworthy.
